### ISSUE TITLE
query regarding how to get started with DS integration and also wrt defaults and overrides

### ISSUE THREAD

**Issue by linxihui** [huggingface/accelerate#1034]
https://github.com/huggingface/accelerate/blob/c0caa068bab67890882d9222011dbdcc0d5e78cd/src/accelerate/accelerator.py#L1076-L1083

As seen above, some config is hardcoded as like `gradient_clipping`. I thought this meant to be some default option, but later on in the code, these configs overwrite user specified configs as by calling `deepspeed_config_process` after (shown below).  So, there is no way to set `gradient_clipping` said to 2. 

https://github.com/huggingface/accelerate/blob/c0caa068bab67890882d9222011dbdcc0d5e78cd/src/accelerate/accelerator.py#L1161

Indeed, `deepspeed_config_process` overwrites many others.  Another issue is that the code does not respect to the non-auto `gradient_accumulation_steps` in the config fie either.  

BTW, the whole config setting is hard to track. There are configs from args, configs from json/yaml,  from ENV. It will be nice to document the priority of each one,  so I know which overwrites which.

**Comment by pacman100** [comment]
This is not an issue, for queries use [Accelerate discussion forum](https://discuss.huggingface.co/c/accelerate/18). Anyways, wrt your queries, please find the responses below:

What you said is incorrect. Quick example:
My config via `accelerate config ` --config_file=ds_zero3.yaml:
```
compute_environment: LOCAL_MACHINE
deepspeed_config:
  gradient_accumulation_steps: 4
  gradient_clipping: 5.8
  offload_optimizer_device: cpu
  offload_param_device: cpu
  zero3_init_flag: true
  zero3_save_16bit_model: true
  zero_stage: 3
distributed_type: DEEPSPEED
downcast_bf16: 'no'
dynamo_backend: 'NO'
fsdp_config: {}
machine_rank: 0
main_training_function: main
megatron_lm_config: {}
mixed_precision: fp16
num_machines: 1
num_processes: 2
rdzv_backend: static
same_network: true
use_cpu: false
```

2. Using `accelerate/examples/nlp_example.py` and making belong changes jsut to see the ds config post `accelerator.prepare` call and then return

![Screenshot 2023-02-01 at 9 13 48 PM](https://user-images.githubusercontent.com/13534540/216090919-1c7fd8cd-2695-4161-a7f1-c4c062aa2c37.png)

3. Output logs:
```
[2023-02-01 16:41:27,487] [INFO] [config.py:997:print_user_config]   json = {
    "train_batch_size": 128, 
    "train_micro_batch_size_per_gpu": 16, 
    "gradient_accumulation_steps": 4, 
    "zero_optimization": {
        "stage": 3, 
        "offload_optimizer": {
            "device": "cpu"
        }, 
        "offload_param": {
            "device": "cpu"
        }, 
        "stage3_gather_16bit_weights_on_model_save": true
    }, 
    "gradient_clipping": 5.8, 
    "steps_per_print": inf, 
    "fp16": {
        "enabled": true, 
        "auto_cast": true
    }, 
    "bf16": {
        "enabled": false
    }, 
    "zero_allow_untested_optimizer": true
}
Using /home/sourab/.cache/torch_extensions/py310_cu117 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0002956390380859375 seconds
{'train_batch_size': 128, 'train_micro_batch_size_per_gpu': 16, 'gradient_accumulation_steps': 4, 'zero_optimization': {'stage': 3, 'offload_optimizer': {'device': 'cpu'}, 'offload_param': {'device': 'cpu'}, 'stage3_gather_16bit_weights_on_model_save': True}, 'gradient_clipping': 5.8, 'steps_per_print': inf, 'fp16': {'enabled': True, 'auto_cast': True}, 'bf16': {'enabled': False}, 'zero_allow_untested_optimizer': True}
```

You can see **gradient_clipping=5.8** and **gradient_accumulation_steps=4** as specified in the `ds_zero3.yaml`

To summarize, the logic you tagged above is for default values and no case of things being hardcoded.

For getting started with DS integration, please go through the following in that order:
1. `DeepSpeed` interactive example in `Launch Configuration` tab of the newly added code explorer: https://huggingface.co/docs/accelerate/usage_guides/explore
2. For most users, answering deepspeed questionnaire of `accelerate config` command is enough and requires no code changes to the source code. This is what I have shown above.
3. For advanced users who want to use the DeepSpeed Config file support along with nuances of using the `auto` fields, refer https://huggingface.co/docs/accelerate/usage_guides/deepspeed#deepspeed-config-file

**Comment by pacman100** [comment]
Could you please change the title to something like `query regarding how to get started with DS integration and also wrt defaults and overrides`

**Comment by linxihui** [comment]
@pacman100 thanks for quick response.  Indeed, I passed DeepSpeed config as an external file, instead of listing inside the accelerate yaml file.

**Comment by linxihui** [comment]
@pacman100  I run with separated deepspeed json config, and it looks ok. I guess I have some Env mess-up. thanks for investigation!
